hal/src/rtl872x/watchdog_hal.cpp

```diff
@@ -0,0 +1,211 @@
+ */
+
+#include "watchdog_hal.h"
+#include "watchdog_base.h"
+
+#if HAL_PLATFORM_HW_WATCHDOG
+#include "check.h"
+#include "logging.h"
+#include "static_recursive_mutex.h"
+
+extern "C" {
+
+#include "rtl8721d.h"
+
+void WDG_IrqClear(void);
+}
+
+class WatchdogLock {
+public:
+    WatchdogLock() {
+        mutex_.lock();
+    }
+
+    ~WatchdogLock() {
+        mutex_.unlock();
+    }
+
+private:
+    StaticRecursiveMutex mutex_;
+};
+
+/**
+ * @note The watchdog will be paused during sleep modes, but not the debug mode.
+ * 
+ */
+class RtlWatchdog : public Watchdog {
+public:
+    int init(const hal_watchdog_config_t* config) {
+        if (started()) {
+            stop();
+        }
+
+        CHECK_TRUE(config && (config->size > 0), SYSTEM_ERROR_INVALID_ARGUMENT);
+        CHECK_TRUE(config->timeout_ms >= WATCHDOG_MIN_TIMEOUT, SYSTEM_ERROR_INVALID_ARGUMENT);
+        CHECK_TRUE(config->timeout_ms <= WATCHDOG_MAX_TIMEOUT, SYSTEM_ERROR_INVALID_ARGUMENT);
+
+        WDG_InitTypeDef WDG_InitStruct = {};
+        uint32_t CountProcess = 0;
+        uint32_t DivFacProcess = 0;
+        WDG_Scalar(config->timeout_ms, &CountProcess, &DivFacProcess);
+        WDG_InitStruct.CountProcess = CountProcess;
+        WDG_InitStruct.DivFacProcess = DivFacProcess;
+        WDG_InitStruct.RstAllPERI = 1;
+        WDG_Init(&WDG_InitStruct);
+        if (config->enable_caps & HAL_WATCHDOG_CAPS_RESET) {
+            BKUP_Set(BKUP_REG0, BIT_KM4SYS_RESET_HAPPEN);
+        } else {
+            // It will change the watchdog to INT mode.
+            WDG_IrqInit((void*)rtlWatchdogEventHandler, (uint32_t)this);
+        }
+        
+        memcpy(&info_.config, config, std::min(info_.config.size, config->size));
+        initialized_ = true;
+        return SYSTEM_ERROR_NONE;
+    }
+
+    int start() {
+        CHECK_TRUE(initialized_, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_FALSE(started(), SYSTEM_ERROR_NONE);
+        WDG_Cmd(ENABLE);
+        CHECK_TRUE(started(), SYSTEM_ERROR_INTERNAL);
+        return SYSTEM_ERROR_NONE;
+    }
+
+    bool started() override {
+        WDG_TypeDef* WDG = ((WDG_TypeDef *) WDG_REG_BASE);
+        uint32_t temp = WDG->VENDOR;
+        if (temp & WDG_BIT_ENABLE) {
+            info_.state = HAL_WATCHDOG_STATE_STARTED;
+        }
+        return info_.state == HAL_WATCHDOG_STATE_STARTED;
+    }
+
+    int stop() override {
+        CHECK_TRUE(started(), SYSTEM_ERROR_NONE);
+        WDG_Cmd(DISABLE);
+        CHECK_FALSE(started(), SYSTEM_ERROR_INTERNAL);
+        return SYSTEM_ERROR_NONE;
+    }
+
+    int refresh() {
+        CHECK_TRUE(initialized_, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_TRUE(started(), SYSTEM_ERROR_INVALID_STATE);
+        WDG_Refresh();
+        return SYSTEM_ERROR_NONE;
+    }
+
+    static RtlWatchdog* instance() {
+        static RtlWatchdog watchdog(HAL_WATCHDOG_CAPS_DEBUG_RUNNING,
+                                    HAL_WATCHDOG_CAPS_RESET | HAL_WATCHDOG_CAPS_NOTIFY_ONLY |
```

Comment:
We can probably do `NOTIFY` + `RESET ` as well if needed, right?

Reply:
No, the watchdog on RTL872x platform supports either reset or INT(notifying).